File: binaryTreeInorderTraversal.js
// 94. Binary Tree Inorder Traversal
// Given the root of a binary tree, return the inorder traversal of its nodes' values.

// /**
//  * Definition for a binary tree node.
//  * function TreeNode(val, left, right) {
//  *     this.val = (val===undefined ? 0 : val)
//  *     this.left = (left===undefined ? null : left)
//  *     this.right = (right===undefined ? null : right)
//  * }
//  */
// /**
//  * @param {TreeNode} root
//  * @return {number[]}
//  */
var inorderTraversalRecursive = function(root) {
  if (!root) return [];
  let res = [];
  
  const helper = function(node) {
      if (!node) return    
      helper(node.left, node);
      res.push(node.val);
      helper(node.right, node);
  }
  
  helper(root);
  return res;
  
};


var inorderTraversalIterative = function(root) {
  if (!root) return [];
  
  let res = [];
  let stack = [];
  let curr = root;
  
  while (curr !== null || stack.length) {
      while (curr !== null) {
          stack.push(curr);
          curr = curr.left;
      }
      
      curr = stack.pop();
      res.push(curr.val);
      curr = curr.right;
      
  }
  
  return res;
  
};

File: binary_tree_inorder_traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class SolutionIterative(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        
        if not root or root == None:
            return []

        stack = []
        res = []
        curr = root
        
        while stack or curr:
            
            while curr:
                stack.append(curr)
                curr = curr.left
            
            curr = stack.pop()
            res.append(curr.val)
            curr = curr.right
        
        return res


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class SolutionRecursive(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        
        res = []
        
        def helper(node):
            if node:
                helper(node.left)
                res.append(node.val)
                helper(node.right)
        
        helper(root)
        
        return res


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class SolutionRecursiveAgain(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        
        if not root:
            return []
        else:
            return self.inorderTraversal(root.left)+[root.val]+self.inorderTraversal(root.right)